# Wrangle

In [27]:
#imports
import pandas as pd
import numpy as np

# system
import sys
# sys.path.append("../data")
sys.path.append("util_")
import wrangle_

In [28]:
# lacd data
df = pd.read_csv("data/known_exploited_vulnerabilities.csv")
df.head()

cveID vendorProject             product  \
0  CVE-2021-27104     Accellion                 FTA   
1  CVE-2021-27102     Accellion                 FTA   
2  CVE-2021-27101     Accellion                 FTA   
3  CVE-2021-27103     Accellion                 FTA   
4  CVE-2021-21017         Adobe  Acrobat and Reader   

                                   vulnerabilityName   dateAdded  \
0   Accellion FTA OS Command Injection Vulnerability  2021-11-03   
1   Accellion FTA OS Command Injection Vulnerability  2021-11-03   
2          Accellion FTA SQL Injection Vulnerability  2021-11-03   
3  Accellion FTA Server-Side Request Forgery (SSR...  2021-11-03   
4  Adobe Acrobat and Reader Heap-based Buffer Ove...  2021-11-03   

                                    shortDescription  \
0  Accellion FTA contains an OS command injection...   
1  Accellion FTA contains an OS command injection...   
2  Accellion FTA contains a SQL injection vulnera...   
3  Accellion FTA contains a server-side request f...   
4  Acrobat Acrobat and Reader contain a heap-base...   

                           requiredAction     dueDate  \
0  Apply updates per vendor instructions.  2021-11-17   
1  Apply updates per vendor instructions.  2021-11-17   
2  Apply updates per vendor instructions.  2021-11-17   
3  Apply updates per vendor instructions.  2021-11-17   
4  Apply updates per vendor instructions.  2021-11-17   

  knownRansomwareCampaignUse notes  cwes  
0                      Known   NaN   NaN  
1                      Known   NaN   NaN  
2                      Known   NaN   NaN  
3                      Known   NaN   NaN  
4                    Unknown   NaN   NaN

In [29]:
# look at the shape
df.shape

(1140, 11)

In [30]:
#look at the columns
df.columns

Index(['cveID', 'vendorProject', 'product', 'vulnerabilityName', 'dateAdded',
       'shortDescription', 'requiredAction', 'dueDate',
       'knownRansomwareCampaignUse', 'notes', 'cwes'],
      dtype='object')

In [31]:
# check the full info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1140 entries, 0 to 1139
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   cveID                       1140 non-null   object 
 1   vendorProject               1140 non-null   object 
 2   product                     1140 non-null   object 
 3   vulnerabilityName           1140 non-null   object 
 4   dateAdded                   1140 non-null   object 
 5   shortDescription            1140 non-null   object 
 6   requiredAction              1140 non-null   object 
 7   dueDate                     1140 non-null   object 
 8   knownRansomwareCampaignUse  1140 non-null   object 
 9   notes                       373 non-null    object 
 10  cwes                        0 non-null      float64
dtypes: float64(1), object(10)
memory usage: 98.1+ KB


In [32]:
# select numeric
len(df.select_dtypes("number").columns) # count of numeric columns

1

In [33]:
#select object columns
len(df.select_dtypes("object").columns) #count of object columns

10

In [34]:
df.describe()

cwes
count   0.0
mean    NaN
std     NaN
min     NaN
25%     NaN
50%     NaN
75%     NaN
max     NaN

**What I see:**


- I have 1140 rows and 11columns
- 1 of the 11 columns are numric while 10 of them are string object colums
- the numeric column `cwes` is null
- the `notes` column has 373 non null rows
- descriptive statistics says nothing on the `cwes` empty column

In [35]:
# remove the non useul unique id and the emplt column
df = df.drop(columns=["cveID", "dateAdded","cwes"])
df.columns

Index(['vendorProject', 'product', 'vulnerabilityName', 'shortDescription',
       'requiredAction', 'dueDate', 'knownRansomwareCampaignUse', 'notes'],
      dtype='object')

In [36]:
# nanipulate the column names to a clear name
spaced_cols = []
col_renamed = ""
for name in df.columns:
    for ele in name:
        if ele == ele.upper():
            col_renamed += " " + ele
        else:
            col_renamed += ele
    spaced_cols.append(col_renamed)
    col_renamed = ""

df.columns = spaced_cols
spaced_cols

['vendor Project',
 'product',
 'vulnerability Name',
 'short Description',
 'required Action',
 'due Date',
 'known Ransomware Campaign Use',
 'notes']

In [37]:
# remove spaces from the ends
# replace " " with _
# convert from lower case
new_cols = df.columns.str.strip().str.replace(" ", "_").str.lower()
df.columns = new_cols #replace the original columns with the new columns
df.columns

Index(['vendor_project', 'product', 'vulnerability_name', 'short_description',
       'required_action', 'due_date', 'known_ransomware_campaign_use',
       'notes'],
      dtype='object')

In [38]:
# remove duplicated rows
print("original data size: ", df.shape)
df = df.drop_duplicates(keep="first")
print("new data size: ", df.shape)

original data size:  (1140, 8)
new data size:  (1099, 8)


In [39]:
# Split the data into 3 sets
train, validate, test = wrangle_.split_data_(df=df, 
                     test_size=0.2, 
                     validate_size= 0.2, 
                     stratify_col="known_ransomware_campaign_use", 
                     random_state=10)
train.shape, validate.shape, test.shape

((659, 8), (220, 8), (220, 8))

In [40]:
# reset index
train = train.reset_index(drop=True)

In [41]:
# count values for each feature
for col in train.columns:
    print(train[col].value_counts)
    print("unique value count size:",train[col].value_counts().shape)
    print("\n\n")

<bound method IndexOpsMixin.value_counts of 0          Zimbra
1          Google
2      SolarWinds
3         Mozilla
4           Rails
          ...    
654         Adobe
655     Hikvision
656     Microsoft
657         Cisco
658        Google
Name: vendor_project, Length: 659, dtype: object>
unique value count size: (121,)



<bound method IndexOpsMixin.value_counts of 0                         Collaboration (ZCS)
1                                 Chromium V8
2                                      Serv-U
3                     Firefox and Thunderbird
4                               Ruby on Rails
                        ...                  
654                                ColdFusion
655               Security cameras web server
656                         Internet Explorer
657    IOS Software and Cisco IOS XE Software
658                               Chromium V8
Name: product, Length: 659, dtype: object>
unique value count size: (308,)



<bound method IndexOpsMixin.value_counts of 0

# Tasks from EDA

* The source of memory corruption is from 4 vendors, but it looks like it's happening on different products of these vendors, so I will go back inthe wrangle file and separate the product from the volunerability and create a additional column for specific vulnerabilities groupd by wording. (cell 7 - cell 15)

In [42]:
train = pd.read_csv("./data/00_temp_train_data.csv", index_col=0)
# create a copy of the training data
temp_train = train.copy()

In [43]:
# # Privilege Escalation Vulnerability  
# search_terms = ["Privilege", "Escalation"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [44]:
# # Memory Corruption Vulnerability
# search_terms = ["Memory", "Corruption"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [45]:
# # Arbitrary File Upload Vulnerability
# search_terms = ["File", "Upload"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [46]:
# # Remote/Arbitrary Code Execution Vulnerability 
# search_terms = ["Execution", "Remote", "Arbitrary"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [47]:
# # Improper Access Control Vulnerability
# search_terms = ["Access", "Improper", "Control"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [48]:
# # Command Injection Vulnerability
# search_terms = ["Injection", "Command", "SQL"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [49]:
# # Use-After-Free Vulnerability 
# search_terms = ["Use-After", "Use-After-Free"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [50]:
# # Deserialization of Untrusted Data Vulnerability
# search_terms = ["Deserialization", "Untrusted"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [51]:
# # Type Confusion Vulnerability  
# search_terms = ["Type Confusion"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [52]:
# # Path Traversal Vulnerability
# search_terms = ["Traversal", "path"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [53]:
# # Buffer Overflow Vulnerability
# search_terms = ["Buffer", "Overflow", "Heap"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [54]:
# # Denial of Service Vulnerability 
# search_terms = ["Denial", "Service"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [55]:
# # Information Disclosure Vulnerability
# search_terms = ["Information", "Disclosure"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [56]:
# # Out-of-Bounds Write Vulnerability
# search_terms = ["Out-of-Bounds", "Write", "Bounds"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [57]:
# # Sandbox Bypass Vulnerability 
# search_terms = ["Sandbox", "Bypass"]
# pattern = '|'.join(search_terms)

# len(train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)])

In [58]:
# # Arbitrary File Upload Vulnerability
# search_terms = ["Unspecified"]
# pattern = '|'.join(search_terms)

# train.vulnerability_name[train['vulnerability_name'].str.contains(pattern, case=False, na=False)]

#### I cam combine all of the obove conditions into one as I engineer a new feature for vulnerabilities

In [59]:
# Create a function that handles feature engineering for vulnerability names
def feature_eng_vul_names(data_frame):
    # Define the conditions
    conditions = [
        data_frame['vulnerability_name'].str.contains(r'Privilege|Escalation', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Memory|Corruption', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'File|Upload', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Remote Code', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Access|Improper|Control', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Injection|Command|SQL', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Use-After|Use-After-Free', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Deserialization|Untrusted', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Type Confusion', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Traversal|path', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Buffer|Overflow|Heap', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Denial|Service', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Information|Disclosure', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Out-of-Bounds|Write|Bounds', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Sandbox|Bypass', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Unspecified', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Spoofing', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Input|Validation', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Hard-Coded|Credentials', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Forgery|SSRF|Request|Server', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Scripting|XSS|Cross|Site', case=False, na=False),
        data_frame['vulnerability_name'].str.contains(r'Arbitrary Code', case=False, na=False)
    ]

    # Define the corresponding values for the new feature
    choices = [
        "Privilege Escalation Vulnerability",
        "Memory Corruption Vulnerability",
        "File Upload Vulnerability",
        "Remote Code Execution Vulnerability",
        "Improper Access Control Vulnerability",
        "Command Injection Vulnerability",
        "Use-After-Free Vulnerability",
        "Deserialization of Untrusted Data Vulnerability",
        "Type Confusion Vulnerability",
        "Path Traversal Vulnerability",
        "Buffer Overflow Vulnerability",
        "Denial of Service Vulnerability",
        "Information Disclosure Vulnerability",
        "Out-of-Bounds Write Vulnerability",
        "Sandbox Bypass Vulnerability",
        "Arbitrary File Upload Vulnerability",
        "Spoofing Vulnerabilities",
        "Input Validation Vulnerabilities",
        "Use of Hard-Coded Credentials Vulnerabilities",
        "Server-Side Request Forgery (SSRF) Vulnerabilities",
        "Cross-Site Scripting (XSS) Vulnerabilities",
        "Arbitrary Code Execution Vulnerability",
    ]

    # Use np.select to create the new feature
    data_frame['vulnerability_type'] = np.select(conditions, choices, default="Other Vulnerability")
    return data_frame


#### The newly engineered feature values

In [60]:
feature_eng_vul_names(train).vulnerability_type.value_counts()

vulnerability_type
Privilege Escalation Vulnerability                    93
Remote Code Execution Vulnerability                   90
Command Injection Vulnerability                       59
Improper Access Control Vulnerability                 46
Sandbox Bypass Vulnerability                          43
Memory Corruption Vulnerability                       39
Buffer Overflow Vulnerability                         39
Use-After-Free Vulnerability                          35
Other Vulnerability                                   34
Path Traversal Vulnerability                          24
Information Disclosure Vulnerability                  23
File Upload Vulnerability                             20
Out-of-Bounds Write Vulnerability                     18
Denial of Service Vulnerability                       18
Deserialization of Untrusted Data Vulnerability       16
Type Confusion Vulnerability                          15
Arbitrary File Upload Vulnerability                   12
Cross-Site S

**Let know look how we can engineer the `required_action` column**

In [118]:
train.required_action.value_counts()

required_action
Apply updates per vendor instructions.                                                                                                                                                                                                                                                                                                                                               505
Apply mitigations per vendor instructions or discontinue use of the product if mitigations are unavailable.                                                                                                                                                                                                                                                                           93
The impacted product is end-of-life and should be disconnected if still in use.                                                                                                                                                       

I see that there is only 15 options that represent the entire training dataset, howerver they are too wordy and I don;t like that, so I will engineer a new column withshort and specific identifiers for each required action .

In [129]:
# Create a function that handles feature engineering for vulnerability names
def feature_eng_req_actions(data_frame):
#     Create an array to contain the required actions
    actions = np.array(data_frame.required_action.value_counts().keys())
    # Define the conditions
    conditions = [
        data_frame['required_action'] == actions[0],
        data_frame['required_action'] == actions[1],
        data_frame['required_action'] == actions[2],
        data_frame['required_action'] == actions[3],
        data_frame['required_action'] == actions[4],
        data_frame['required_action'] == actions[5],
        data_frame['required_action'] == actions[6],
        data_frame['required_action'] == actions[7],
        data_frame['required_action'] == actions[8],
        data_frame['required_action'] == actions[9],
        data_frame['required_action'] == actions[10],
        data_frame['required_action'] == actions[11],
        data_frame['required_action'] == actions[12],
        data_frame['required_action'] == actions[13],
        data_frame['required_action'] == actions[14]
    ]

    # Define the corresponding values for the new feature
    choices = [
        "Apply Updates", 
        "Apply Mitigations or Discontinue", 
        "End of Life Disconnect",  
        "Apply Updates or Discontinue",  
        "Apply Remediations or Discontinue",  
        "Apply Updates or Discontinue",  
        "Legacy Disconnect",  
        "Verify Cisco Compliance",  
        "Mitigate GroupLock or Discontinue",  
        "End of Life Disconnect",  
        "Mitigate or Disable SLP",  
        "Log4j Remediation",  
        "Internet Block and Update",  
        "Enable Threat Prevention",  
        "DLink Update or Disconnect",
    ]

    # Use np.select to create the new feature
    data_frame['required_action_new'] = np.select(conditions, choices)
    return data_frame


In [130]:
feature_eng_req_actions(train).required_action_new.value_counts()

required_action_new
Apply Updates                        505
Apply Mitigations or Discontinue      93
End of Life Disconnect                32
Apply Updates or Discontinue          12
Apply Remediations or Discontinue      6
Legacy Disconnect                      3
Verify Cisco Compliance                2
Mitigate GroupLock or Discontinue      1
Mitigate or Disable SLP                1
Log4j Remediation                      1
Internet Block and Update              1
Enable Threat Prevention               1
DLink Update or Disconnect             1
Name: count, dtype: int64

In [131]:
train.required_action.value_counts().values

array([505,  93,  31,   8,   6,   4,   3,   2,   1,   1,   1,   1,   1,
         1,   1])

**Apply new changes to the entire dataframe**

In [62]:
df = feature_eng_vul_names(df)
df.head(3)

vendor_project product                                vulnerability_name  \
0      Accellion     FTA  Accellion FTA OS Command Injection Vulnerability   
1      Accellion     FTA  Accellion FTA OS Command Injection Vulnerability   
2      Accellion     FTA         Accellion FTA SQL Injection Vulnerability   

                                   short_description  \
0  Accellion FTA contains an OS command injection...   
1  Accellion FTA contains an OS command injection...   
2  Accellion FTA contains a SQL injection vulnera...   

                          required_action    due_date  \
0  Apply updates per vendor instructions.  2021-11-17   
1  Apply updates per vendor instructions.  2021-11-17   
2  Apply updates per vendor instructions.  2021-11-17   

  known_ransomware_campaign_use notes               vulnerability_type  
0                         Known   NaN  Command Injection Vulnerability  
1                         Known   NaN  Command Injection Vulnerability  
2                         Known   NaN  Command Injection Vulnerability

In [64]:
# remove columns not needed
df = df.drop(columns=["notes", "short_description", "vulnerability_name", "due_date"])
df.head(3)

vendor_project product                         required_action  \
0      Accellion     FTA  Apply updates per vendor instructions.   
1      Accellion     FTA  Apply updates per vendor instructions.   
2      Accellion     FTA  Apply updates per vendor instructions.   

  known_ransomware_campaign_use               vulnerability_type  
0                         Known  Command Injection Vulnerability  
1                         Known  Command Injection Vulnerability  
2                         Known  Command Injection Vulnerability

### Dummies

Create dummy variables for modeling

In [78]:
# First create a copy of the original dataframe
df_dummy = df.copy()

In [81]:
# get all the columns to upply dummies to
dummy_col = df_dummy.columns[df_dummy.columns != "known_ransomware_campaign_use"]

df_dummy[dummy_col].head(3)

vendor_project                                 product  \
0         Accellion                                     FTA   
1         Accellion                                     FTA   
2         Accellion                                     FTA   
3         Accellion                                     FTA   
4             Adobe                      Acrobat and Reader   
...             ...                                     ...   
1135      Microsoft                       Internet Explorer   
1136        Acronis              Cyber Infrastructure (ACI)   
1137     ServiceNow  Utah, Vancouver, and Washington DC Now   
1138     ServiceNow  Utah, Vancouver, and Washington DC Now   
1139         VMware                                    ESXi   

                                        required_action  \
0                Apply updates per vendor instructions.   
1                Apply updates per vendor instructions.   
2                Apply updates per vendor instructions.   
3                Apply updates per vendor instructions.   
4                Apply updates per vendor instructions.   
...                                                 ...   
1135  The impacted product is end-of-life and should...   
1136  Apply mitigations per vendor instructions or d...   
1137  Apply mitigations per vendor instructions or d...   
1138  Apply mitigations per vendor instructions or d...   
1139  Apply mitigations per vendor instructions or d...   

                                     vulnerability_type  
0                       Command Injection Vulnerability  
1                       Command Injection Vulnerability  
2                       Command Injection Vulnerability  
3     Server-Side Request Forgery (SSRF) Vulnerabili...  
4                         Buffer Overflow Vulnerability  
...                                                 ...  
1135                       Use-After-Free Vulnerability  
1136                                Other Vulnerability  
1137                    Denial of Service Vulnerability  
1138              Improper Access Control Vulnerability  
1139                       Sandbox Bypass Vulnerability  

[1099 rows x 4 columns]

**Make the targer binary**

* All `known` ransomware are labled as `1`, and all `unknown` copains are labeled with `0`.

In [87]:
df_dummy.known_ransomware_campaign_use

df_dummy["ransomware_use"] = np.where(df_dummy["known_ransomware_campaign_use"] == "Known", 1, 0)
df_dummy.head(3)

vendor_project product                         required_action  \
0      Accellion     FTA  Apply updates per vendor instructions.   
1      Accellion     FTA  Apply updates per vendor instructions.   
2      Accellion     FTA  Apply updates per vendor instructions.   

  known_ransomware_campaign_use               vulnerability_type  \
0                         Known  Command Injection Vulnerability   
1                         Known  Command Injection Vulnerability   
2                         Known  Command Injection Vulnerability   

   ransomware_use  
0               1  
1               1  
2               1

## Split and Save all the data sets

In [65]:
# Split the data into 3 sets
train, validate, test = wrangle_.split_data_(df=df, 
                     test_size=0.2, 
                     validate_size= 0.2, 
                     stratify_col="known_ransomware_campaign_use", 
                     random_state=10)
train.shape, validate.shape, test.shape

((659, 5), (220, 5), (220, 5))

In [66]:
known_exploited_vulnerabilities.csv

NameError: name 'known_exploited_vulnerabilities' is not defined